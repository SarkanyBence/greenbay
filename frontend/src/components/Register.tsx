import { useEffect, useState } from "react";
import { useHistory } from "react-router";
import fetchData from "../services/fetchData";
import NewUser from "../types/newUser";
import NewUserDto from "../types/newUserDto";
import PropsInput from "../types/PropsInput";
import FormInput from "./FormInput";

function Register() {
  const history = useHistory();
  const [userState, setUserState] = useState({
    userName: "",
    email: "",
    password: "",
    confirmPassword: "",
  } as NewUser);
  const [userError, setUserError] = useState(false);

  useEffect(() => {}, [userError]);

  const submitForm = (e: any) => {
    e.preventDefault();
    const data = Object.fromEntries(new FormData(e.target).entries());
    const dataDto: NewUserDto = new NewUserDto(data as unknown as NewUser);
    console.log(userState);
    fetchData("POST", "/register", dataDto)
      .then((res) => {
        history.push("/login");
      })
      .catch((err) => {
        if (err.status === 422) {
          setUserError(true);
        }
      });
  };

  const handleChange = (e: any) => {
    if (e.target.name === "userName") setUserError(false);
    setUserState({ ...userState, [e.target.name]: e.target.value });
  };

  const inputs: PropsInput[] = [
    {
      id: 1,
      name: "userName",
      placeholder: "Username",
      type: "text",
      errorMessage:
        "Username should be 3-16 characters, no special characters allowed!",
      pattern: "^[A-Za-z0-9]{3,16}$",
      required: true,
    },
    {
      id: 2,
      name: "email",
      placeholder: "Email",
      type: "email",
      errorMessage: "It should be a valid email address!",
      required: true,
    },
    {
      id: 3,
      name: "password",
      placeholder: "Password",
      type: "password",
      errorMessage: "Password should be 8-20 characters!",
      pattern: ".{8,20}",
      required: true,
    },
    {
      id: 4,
      name: "confirmPassword",
      placeholder: "Confirm password",
      type: "password",
      errorMessage: "Passwords don't match",
      pattern: userState.password,
      required: true,
    },
  ];

  return (
    <div className="base-container">
      <form onSubmit={submitForm}>
        <span
          className="formInput other-error"
          style={{ opacity: userError ? 1 : 0 }}
        >
          Username is already taken!
        </span>
        {inputs.map((input) => (
          <FormInput
            key={input.id}
            {...input}
            userState={userState}
            handleChange={handleChange}
          />
        ))}

        <button>Sign Up</button>
      </form>
    </div>
  );
}

export default Register;
