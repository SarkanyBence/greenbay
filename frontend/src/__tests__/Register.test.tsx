import Register from "../components/Register";
import * as ReactDom from "react-dom";
import renderer from "react-test-renderer";

describe("Register tests", () => {
  let container: HTMLDivElement;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    ReactDom.render(<Register />, container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container.remove();
  });

  it("Register snapshot test", () => {
    const tree = renderer.create(<Register />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.skip ("renders correctly", () => {
    const inputs = document.getElementsByName("input");
    expect(inputs.length).toBe(4);
  });
});