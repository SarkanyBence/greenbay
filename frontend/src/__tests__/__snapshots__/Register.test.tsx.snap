// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register tests Register snapshot test 1`] = `
<div
  className="base-container"
>
  <form
    onSubmit={[Function]}
  >
    <span
      className="formInput other-error"
      style={
        Object {
          "opacity": 0,
        }
      }
    >
      Username is already taken!
    </span>
    <div
      className="formInput"
    >
      <input
        data-focused="false"
        name="userName"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        pattern="^[A-Za-z0-9]{3,16}$"
        placeholder="Username"
        required={true}
        type="text"
        value=""
      />
      <span
        className="formInput error"
      >
        Username should be 3-16 characters, no special characters allowed!
      </span>
    </div>
    <div
      className="formInput"
    >
      <input
        data-focused="false"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Email"
        required={true}
        type="email"
        value=""
      />
      <span
        className="formInput error"
      >
        It should be a valid email address!
      </span>
    </div>
    <div
      className="formInput"
    >
      <input
        data-focused="false"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        pattern=".{8,20}"
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
      <span
        className="formInput error"
      >
        Password should be 8-20 characters!
      </span>
    </div>
    <div
      className="formInput"
    >
      <input
        data-focused="false"
        name="confirmPassword"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        pattern=""
        placeholder="Confirm password"
        required={true}
        type="password"
        value=""
      />
      <span
        className="formInput error"
      >
        Passwords don't match
      </span>
    </div>
    <button>
      Sign Up
    </button>
  </form>
</div>
`;
